services/venice-router/src/main/java/com/linkedin/venice/router/api/path/VeniceComputePath.java

```diff
@@ -22,30 +22,44 @@
 import com.linkedin.venice.serializer.RecordSerializer;
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.BiConsumer;
 import javax.annotation.Nonnull;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
 import org.apache.avro.io.BinaryDecoder;
 import org.apache.avro.io.OptimizedBinaryDecoderFactory;
 
 
 public class VeniceComputePath extends VeniceMultiKeyPath<ComputeRouterRequestKeyV1> {
+  private static final Schema EMPTY_RECORD_SCHEMA = Schema.createRecord(
+      ComputeRequestV3.class.getSimpleName(),
+      "no-op",
+      ComputeRequestV3.class.getPackage().getName(),
```

Comment:
With this logic, if v4 becomes a wire protocol or we add a v5, which contains some new optional field, we need to change `ComputeRequestV3` to the latest protocol, right?
If v5 adds a new optional field instead of adding a branch in the union types, we will need to use the wire protocol version to deserialize it rather than the latest compute version: v5, otherwise, the deserialization won't work from my understanding.
Can you clarify?
Can you clarify?

Reply:
We must always know the correct writer schema (which is the one the client used to encode its request). Previously and in this commit too we have the assumption that the backend knows all protocol versions the client might use, and if that assumption is violated, it's ok to fail the request.

Making new clients compatible with older backends is not supported (neither before nor in this commit).

Reply:
I may have a misunderstanding here.. 
```
  private static final RecordDeserializer<GenericRecord> COMPUTE_REQUEST_NO_OP_DESERIALIZER =
      FastSerializerDeserializerFactory.getFastAvroGenericDeserializer(ComputeRequestV3.SCHEMA$, EMPTY_RECORD_SCHEMA);
```
So the writer schema is always v3, and it will certainly work now since all other versions are not being used as wire protocol.
In the future, I assume we will change this logic when we add more wire protocol versions, right?

Reply:
Yes, that would need to change. Depending on how the protocol is evolved, the change here could differ, e.g.:

1. If we add more branches to the same union we kept evolving so far, then this writer schema can be bumped to the latest and that's it. A single schema could still deserialize all versions.

2. If we we add fields elsewhere in the schema, then we would need two deserializers, this old one we already have plus a new one, and we would need to pick the right one based on the API version header.

Reply:
SG.
Can you add some javadoc about ^?
It is a little confusing when first read this piece of code, and also, the dev of read compute needs to be aware of this when evolving read compute schemas.

Reply:
Will do. I can't get to it today but I'll update the PR on Monday. Thanks for the reviews!

Reply:
Done